Remove component definitions from package __init__ file

gonzalo-bulnes opened the pull request:
# Description

This is the first step to be split out of #1369.

The presence of component definitions in the `gui/__init__.py` file is a source of circular dependencies when trying to split out `gui/widgets.py`.

This PR moves those components to a `misc` package (until it becomes clear where they fit best), completes test coverage (the module wasn't entirely covered by its own test file) and propose an (optional) rewrite of the module tests that ensures the _behavior_ of the components is tests independently from their implementation.

**Note**: the API of the `gui` package doesn't change, only its internal layout.

# Test Plan

_I recommend reviewing this PR **commit by commit**._

The first commit is a refactoring:
- [ ] Confirm that the components haven't changed.
- [ ] Confirm that the tests haven't changed
- [ ] Confirm that the build is green

The following 4 commits are an optional test suite enhancement:
- [ ] Confirm that the tests are were removed were adequately replaced
- [ ] Confirm that the additional tests make sense
- [ ] Confirm that the module tests ensure proper coverage
      `pytest tests/gui/test_misc.py -v --cov-config .coveragerc --cov-report html --cov-report term-missing --cov=securedrop_client.gui.misc --cov-fail-under 100`
- [ ] Confirm that the build is green

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [x] These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - [ ] I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [x] No update to the AppArmor profile is required for these changes
 - [ ] I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [ ] I have written a migration and upgraded a test database based on `main` and confirmed that the migration applies cleanly
 - [ ] I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - [ ] I need help writing a database migration
 - [x] No database schema changes are needed

gonzalo-bulnes commented:
@creviera These components don't have stylesheets. How does it look to you? (what do you think of the tests reshaping? Those could go out of the PR, but honestly if we're touching test files I'd ensure they cover the modules and do it by testing the components behavior as much as possible. I say "as much as possible" because sometimes that requires refactoring and that's where I'd draw the line for now: not modifying the code while we move it around.)

gonzalo-bulnes commented:
@creviera Can you run the test suite locally and tell me if it passes? I triggered this last build twice, and it failed consistently but locally if behaves as expected (fails when I modify the `max_length` because the elision moves slightly just as I'd expect). I'm not sure why it would be failing in CI but not locally :thinking:

creviera commented on securedrop_client/gui/__init__.py:
```diff
@@ -16,193 +16,11 @@
-
-# Allow the buttons to be imported directly.
+# Expose GUI elements without exposing the package structure
 from securedrop_client.gui.buttons import SDPushButton  # noqa: F401
```
why have an unused import here? The comment to expose gui elements without exposing package structure may confuse future devs. I think we should just remove this line.

gonzalo-bulnes replied:
This import is what allows to use `SDPushButton` from `securedrop_client.gui`, the `__init__.py` file is the file that defines the API of the package. In this case, as far as a consumer is concerned, the package defines an `SDPushButton` class and the fact that the package is internally split into multiple files is an implementation detail.

None of the imports present in this file will be used _in the file itself_, but they're used in other files, and the linter is overzealous in this case.

You can look at the `__init__.py` files in #1369 and #1374 for further examples of how the API of `securedrop_client.gui` is built from its internal packages. 

The goal is to distinguish between what's the public API of each package and what is private. Eventually imports shouldn't need to be _deeply_ nested (`from securedrop_client.gui import SourceList` vs `from securedrop_client.gui.source.list import List`) because the consumer shouldn't be coupled to the internal structure of a package. (It is the package version of [Demeter's guideline][wiki]: too many "dots" in an import are a smell.)

  [wiki]: https://en.m.wikipedia.org/wiki/Law_of_Demeter

I introduced relative import paths in #1374, that you'll see help making the distinction visually between imports that _define_ APIs (in `__init__.py_ files, should all be relative) and consumer imports which happen outside of `__init__.py` files and refer to sibling packages (e.g. when `securedrop_client.logic` is referenced from `securedrop_client.gui`). Currently `securedrop_client.gui` etc. is just repeated everywhere and obscures the package hierarchy a bit.

That's a bit of a long explanation, am I making sense?

gonzalo-bulnes replied:
The comment was attempting to explain why the import was not removed, even though the linter would complain that it was (apparently) not used! That didn't work! 😝

creviera replied:
Okay, the change makes sense. Let's update this comment `Expose GUI elements without exposing the package structure` to something more prescriptive, like, "Put imports you want accessible from securedrop_client.gui here."

gonzalo-bulnes commented:
> @creviera @rocodes Could you please run the test suite locally and tell me if it passes? (make test or pytest tests/gui/test_misc.py -v should do it.) I triggered this last build twice, and it failed consistently but locally if behaves as expected (fails when I modify the max_length because the elision moves slightly just as I'd expect). I'm not sure why it would be failing in CI but not locally :thinking: 

I modified the params slightly and it's green now. :green_apple:  I believe (pretty sure) the `80px` limit would just require elision or not depending on small variations in rendering, specially since the last character is narrow (`!`). (Here is the [previously failing test case][ref] for future reference.)
I picked the new values making sure there is margin either way. I think there isn't much way around this given we're dealing with pixel values and font rendering, so I'm not worries about it in general, and specifically I don't expect that test to be flaky going forward.

  [ref]: https://app.circleci.com/pipelines/github/freedomofpress/securedrop-client/1317/workflows/7b962370-8333-44e5-9bc2-7043a7e33d88/jobs/6728/tests#failed-test-0

creviera commented on securedrop_client/gui/misc.py:
```diff
@@ -0,0 +1,207 @@
+        layout.setSpacing(0)
+
+        # Add SVG icon and set its size
+        self._icon = load_icon(normal=on, normal_off=off)
```
note: this class was copied exactly except for adding the `_` here and with other private members.

gonzalo-bulnes replied:
I removed this from the commit. It really did belong in the commit after when using `button.icon()` (part of the `QPushButton` API) in the test file.

creviera commented on tests/gui/test_misc.py:
```diff
@@ -85,8 +85,8 @@ def test_SvgLabel_init(mocker):
-    load_svg_fn = mocker.patch("securedrop_client.gui.load_svg", return_value=svg)
-    mocker.patch("securedrop_client.gui.QHBoxLayout.addWidget")
+    load_svg_fn = mocker.patch("securedrop_client.gui.misc.load_svg", return_value=svg)
+    mocker.patch("PyQt5.QtWidgets.QHBoxLayout.addWidget")
```
gorg!

creviera commented:
Due to time, the first commit lgtm and is ready to be merged (with the one nit of adding a clearer code comment and with _just_ a test move (not a rewrite). The rewrite of tests can be decoupled from the "Download All" work.

creviera commented:
also, this can be based against `main`

gonzalo-bulnes commented:
(rebased)

- Removed the `icon` -> `_icon` that didn't belong in the commit where it was staged
- Removed the (optional) test re-write, I'll add them as a follow-up PR in a way that doesn't remove (only marks for review/removal) the existing tests.

creviera reviewed (dismissed):
lgtm

creviera commented:
waiting on ci...
